core/src/main/java/io/questdb/cairo/DatabaseSnapshotAgentImpl.java

```diff
@@ -0,0 +1,487 @@
+                throw CairoException.critical(ff.errno()).put("Could not create [dir=").put(path).put(']');
+            }
+
+            // Suspend the WalPurgeJob
+            if (walPurgeJobRunLock != null) {
+                final long timeout = configuration.getCircuitBreakerConfiguration().getTimeout();
+                while (!walPurgeJobRunLock.tryLock(timeout, TimeUnit.MICROSECONDS)) {
+                    executionContext.getCircuitBreaker().statefulThrowExceptionIfTrippedNoThrottle();
+                }
+            }
+
+            try {
+                // Prepare table name registry for copying.
+                path.trimTo(snapshotDbLen).$();
+                tableNameRegistryStore.of(path, 0);
+
+                path.trimTo(snapshotDbLen).$();
+                try (
+                        TableListRecordCursorFactory factory = new TableListRecordCursorFactory();
+                        RecordCursor cursor = factory.getCursor(executionContext);
+                        MemoryCMARW mem = Vm.getCMARWInstance()
+                ) {
+                    final int tableNameIndex = factory.getMetadata().getColumnIndex(TableListRecordCursorFactory.TABLE_NAME_COLUMN);
+                    final Record record = cursor.getRecord();
+
+                    // Copy metadata files for all tables.
+                    while (cursor.hasNext()) {
+                        CharSequence tableName = record.getStr(tableNameIndex);
+                        path.of(configuration.getRoot());
+                        TableToken tableToken = engine.verifyTableName(tableName);
+                        if (
+                                TableUtils.isValidTableName(tableName, tableName.length())
+                                        && ff.exists(path.concat(tableToken).concat(TableUtils.META_FILE_NAME).$())
+                        ) {
+                            boolean isWalTable = engine.isWalTable(tableToken);
+                            path.of(configuration.getSnapshotRoot()).concat(configuration.getDbDirectory());
+                            LOG.info().$("preparing for snapshot [table=").$(tableName).I$();
+
+                            path.trimTo(snapshotDbLen).concat(tableToken);
+                            int rootLen = path.length();
+                            if (isWalTable) {
+                                path.concat(WalUtils.SEQ_DIR);
+                            }
+                            if (ff.mkdirs(path.slash$(), configuration.getMkDirMode()) != 0) {
+                                throw CairoException.critical(ff.errno()).put("could not create [dir=").put(path).put(']');
+                            }
+
+                            try (TableReader reader = engine.getReaderWithRepair(tableToken)) {
+                                // Add entry to table name registry copy.
+                                tableNameRegistryStore.logAddTable(tableToken);
```

Comment:
Good point. It's a pity we don't have all tables there though as it may be handy for further snapshot hardening. Added a check in d05ac4c6e931e5bdb1594771d6be7e64174db15d